src/sdc11073/mdib/providermdib.py

```diff
@@ -0,0 +1,182 @@
+from typing import TYPE_CHECKING, Any
+
+from sdc11073 import loghelper
+from sdc11073.definitions_base import ProtocolsRegistry
+from sdc11073.definitions_sdc import SDC_v1_Definitions
+from sdc11073.observableproperties import ObservableProperty
+from sdc11073.pysoap.msgreader import MessageReader
+
+from . import mdibbase
+from .providermdibxtra import ProviderMdibMethods
+from .transactions import MdibUpdateTransaction, RtDataMdibUpdateTransaction
+
+if TYPE_CHECKING:
+    from sdc11073.definitions_base import BaseDefinitions
+
+    from .transactions import TransactionManagerProtocol
+
+
+class ProviderMdib(mdibbase.MdibBase):
+    """Device side implementation of a mdib.
+
+    Do not modify containers directly, use transactions for that purpose.
+    Transactions keep track of changes and initiate sending of update notifications to clients.
+    """
+
+    transaction = ObservableProperty(fire_only_on_changed_value=False)
+    rt_updates = ObservableProperty(fire_only_on_changed_value=False)  # different observable for performance
+
+    def __init__(self,
+                 sdc_definitions: type[BaseDefinitions] | None = None,
+                 log_prefix: str | None = None,
+                 extra_functionality: type | None = None,
+                 transaction_cls: type[TransactionManagerProtocol] | None = None,
+                 ):
+        """Construct a ProviderMdib.
+
+        :param sdc_definitions: defaults to sdc11073.definitions_sdc.SDC_v1_Definitions
+        :param log_prefix: a string
+        :param extra_functionality: class for extra functionality, default is ProviderMdibMethods
+        :param transaction_cls: runs the transaction, default is MdibUpdateTransaction
+        """
+        if sdc_definitions is None:
+            sdc_definitions = SDC_v1_Definitions
+        super().__init__(sdc_definitions,
+                         loghelper.get_logger_adapter('sdc.device.mdib', log_prefix))
+        if extra_functionality is None:
+            extra_functionality = ProviderMdibMethods
+        self._xtra = extra_functionality(self)
+        self._tr_lock = Lock()  # transaction lock
+
+        self.sequence_id = uuid.uuid4().urn  # this uuid identifies this mdib instance
+
+        self._annotators = {}
+        self._current_transaction = None
+
+        self.pre_commit_handler = None  # pre_commit_handler can modify transaction if needed before it is committed
+        self.post_commit_handler = None  # post_commit_handler can modify mdib if needed after it is committed
+        self._transaction_cls = transaction_cls or MdibUpdateTransaction
+        self._retrievability_episodic = []  # a list of handles
+        self.retrievability_periodic = defaultdict(list)
+
+    @property
+    def xtra(self) -> Any:
+        """Give access to extended functionality."""
+        return self._xtra
+
+    @contextmanager
+    def transaction_manager(self, set_determination_time: bool = True) -> TransactionManagerProtocol:
```

Comment:
correct typing is `typing.ContextManager[TransactionManagerProtocol]`

Reply:
ok